great_expectations/data_context/data_context/abstract_data_context.py

```diff
@@ -461,6 +461,42 @@ def set_config(self, project_config: DataContextConfig) -> None:
         self._project_config = project_config
         self.variables.config = project_config
 
+    def save_datasource(
+        self, datasource: Union[LegacyDatasource, BaseDatasource]
+    ) -> Union[LegacyDatasource, BaseDatasource]:
+        """Save a Datasource to the configured DatasourceStore.
+
+        Stores the underlying DatasourceConfig in the store, updates the cached Datasource and returns the Datasource.
+        The cached and returned Datasource is re-constructed from the config that was stored as some store implementations
+        make edits to the stored config (e.g. adding identifiers).
+
+        Args:
+            datasource: Datasource to store.
+
+        Returns:
+            The datasource, after storing and retrieving the stored config.
+        """
+
+        datasource_config_dict: dict = datasourceConfigSchema.dump(datasource.config)
+        datasource_config = DatasourceConfig(**datasource_config_dict)
+        datasource_name: str = datasource.name
+
+        updated_datasource_config_from_store: DatasourceConfig = self._datasource_store.set(  # type: ignore[attr-defined]
+            key=None, value=datasource_config
+        )
+        # Use the updated datasource config, since the store may populate additional info on update.
+        self.config.datasources[datasource_name] = updated_datasource_config_from_store  # type: ignore[index,assignment]
+
+        # Also use the updated config to initialize a datasource for the cache and overwrite the existing datasource.
+        substituted_config = self._perform_substitutions_on_datasource_config(
+            updated_datasource_config_from_store
+        )
+        updated_datasource: Datasource = self._instantiate_datasource_from_config(
```

Comment:
I thought we explored the idea of creating a new datasource and returning it but then decided that that might be confusing because there might be references to the datasource elsewhere (eg in a validator) and we might need to guard against using the old, "dirty" datasource. I thought we settled on just updating in place. Does the line above update in place and, if so, can we do that instead of creating a new datasource?

Reply:
Good call. I was going to work on that separately, but since this PR is going in after the release it makes sense to do that in the current PR. I'll make that change and add some related tests.

Reply:
On further discussion with @donaldheppner we decided to forego that change in the current PR.